### DISCUSSION TITLE
Using box shadow color with custom box shadow appears to strip css variables

### DISCUSSION THREAD

**Discussion by CodeWithOz** [tailwindlabs/tailwindcss#9821] [Help]
I'm using tailwind in nextjs and i've added a css variable as follows:
```css
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
    :root {
        --shadow-panel-height: 20px;
    }
}
```

In my tailwind config I add a custom box shadow that uses that css variable, as follows:
```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  ...
  theme: {
    extend: {
      boxShadow: {
        'inner-2': 'inset 0 var(--shadow-panel-height) 20px 0 rgb(0 0 0 / 0.05);'
      }
    },
  },
  ...
}
```

Now I want to update the color of the box shadow using tailwind's `shadow-[color]` convention. So I use a class like `shadow-inner-2 shadow-amber-900` but that causes my variable to stop being used. I noticed in dev tools that the `--tw-shadow-colored` variable doesn't include my variable, even though the `--tw-shadow` variable does. This is a screenshot:

<img width="310" alt="Screen Shot 2022-11-13 at 10 32 24 AM" src="https://user-images.githubusercontent.com/28525986/201515220-6823cad5-c87c-4d06-ac80-400ea0f1905b.png">

It's subtle but it means that when I use only `shadow-inner-2` then my variable is used and the shadow looks like I expect. But if I then try to customize the color by combining with `shadow-[color]` then the shadow is different from what I expect because the variable isn't used.  This looks like a bug to me, but perhaps I'm missing something. Thoughts?

**Comment by CodeWithOz** [comment]
For now the temporary workaround I'm using is to expose all tailwind colors as css variables using [this plugin](https://gist.github.com/Merott/d2a19b32db07565e94f10d13d11a8574) and then using the color I want directly in the custom box shadow. So my tailwind config looks like this:
```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  ...
  theme: {
    extend: {
      boxShadow: {
        // notice the use of var(--color-amber-900) now
        'inner-2': 'inset 0 var(--shadow-panel-height) 10px 0 var(--color-amber-900);'
      }
    },
  },
  plugins: [
    // lovely snippet to export all tailwind colors as css variables
    // source: https://gist.github.com/Merott/d2a19b32db07565e94f10d13d11a8574
    function({ addBase, theme }) {
      function extractColorVars(colorObj, colorGroup = '') {
        return Object.keys(colorObj).reduce((vars, colorKey) => {
          const value = colorObj[colorKey];

          const newVars =
            typeof value === 'string'
              ? { [`--color${colorGroup}-${colorKey}`]: value }
              : extractColorVars(value, `-${colorKey}`);

          return { ...vars, ...newVars };
        }, {});
      }

      addBase({
        ':root': extractColorVars(theme('colors')),
      });
    },
  ],
}
```

Of course I still can't customize the color of the shadow dynamically using `shadow-[color]` classes, I can only define the color directly in the theme config.

**Reply by wongjn** [reply]
Another work-around method without having to export all colors could be:
```diff
 /** @type {import('tailwindcss').Config} */
 module.exports = {
   …
   theme: {
     extend: {
       boxShadow: {
-       'inner-2': 'inset 0 var(--shadow-panel-height) 20px 0 rgb(0 0 0 / 0.05);'
+       'inner-2': 'inset 0 var(--shadow-panel-height) 20px 0 var(--shadow-color,rgb(0 0 0 / 0.05));'
       }
     },
   },
   …
 }
```
So it still uses the default translucent black, but if you wanted to use a custom color, you could use an arbitrary property instead of `shadow-amber-900`, `[--shadow-color:theme(colors.amber.900)]`. And you could convert this into a custom plugin if you so wished.

**Reply by CodeWithOz** [reply]
🤯🔥 love that! Much better solution, thanks 🙌🏾.

**Comment by wongjn** [comment]
It seems like box shadow value parsing does not take into account CSS variables:
https://github.com/tailwindlabs/tailwindcss/blob/f2f1ee9b5585d11945587f688a4c533c5287aaf7/src/util/parseBoxShadowValue.js#L5
https://github.com/tailwindlabs/tailwindcss/blob/f2f1ee9b5585d11945587f688a4c533c5287aaf7/src/util/parseBoxShadowValue.js#L26-L40
